File: app/src/main/java/com/tunjid/androidbootstrap/viewholders/TileViewHolder.java
package com.tunjid.androidbootstrap.viewholders;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.graphics.Color;
import android.view.View;
import android.widget.TextView;

import com.tunjid.androidbootstrap.R;
import com.tunjid.androidbootstrap.adapters.TileAdapter;
import com.tunjid.androidbootstrap.model.Tile;
import com.tunjid.androidbootstrap.recyclerview.InteractiveViewHolder;

public class TileViewHolder extends InteractiveViewHolder<TileAdapter.AdapterListener> {

    private static final int COLOR_CHANGE_DURATION = 1000;
    private static final int START_DELAY = 300;

    private TextView text;
    private Tile tile;

    private final ValueAnimator animator;
    private final ValueAnimator.AnimatorUpdateListener listener;

    public TileViewHolder(View itemView, TileAdapter.AdapterListener scanAdapterListener) {
        super(itemView, scanAdapterListener);

        text = itemView.findViewById(R.id.tile_text);
        animator = ValueAnimator.ofObject(new ArgbEvaluator(), Color.RED);
        animator.setDuration(COLOR_CHANGE_DURATION);

        listener = this::updateTextColor;
        itemView.setOnClickListener(view -> adapterListener.onTileClicked(tile));
    }

    public void bind(Tile tile) {
        this.tile = tile;
        text.setText(tile.getId());

        animator.setIntValues(text.getCurrentTextColor(), tile.getColor());
        animator.addUpdateListener(listener);
        animator.setStartDelay(START_DELAY); // Cheeky bit of code to keep scrolling smooth on fling
        animator.start();
    }

    public void unBind() {
        animator.cancel();
        animator.removeUpdateListener(listener);
    }

    private void updateTextColor(ValueAnimator animation) {
        text.setTextColor((int) animation.getAnimatedValue());
    }
}

File: app/src/main/java/com/tunjid/androidbootstrap/viewholders/TileViewHolder.kt
package com.tunjid.androidbootstrap.viewholders

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.graphics.Color
import android.view.View
import android.widget.TextView
import com.tunjid.androidbootstrap.R
import com.tunjid.androidbootstrap.model.Tile
import com.tunjid.androidbootstrap.recyclerview.InteractiveViewHolder

class TileViewHolder(itemView: View, scanAdapterListener: (tile: Tile) -> Unit)
    : InteractiveViewHolder<(tile: Tile) -> Unit>(itemView, scanAdapterListener) {

    private lateinit var tile: Tile
    private val text: TextView = itemView.findViewById(R.id.tile_text)
    private val animator: ValueAnimator = ValueAnimator.ofObject(ArgbEvaluator(), Color.RED)
    private val listener: ValueAnimator.AnimatorUpdateListener = ValueAnimator.AnimatorUpdateListener(this::updateTextColor)

    init {
        animator.duration = COLOR_CHANGE_DURATION.toLong()
        itemView.setOnClickListener { delegate.invoke(tile) }
    }

    fun bind(tile: Tile) {
        this.tile = tile
        text.text = tile.id

        animator.setIntValues(text.currentTextColor, tile.color)
        animator.addUpdateListener(listener)
        animator.startDelay = START_DELAY.toLong() // Cheeky bit of code to keep scrolling smooth on fling
        animator.start()
    }

    fun unBind() {
        animator.cancel()
        animator.removeUpdateListener(listener)
    }

    private fun updateTextColor(animation: ValueAnimator) {
        text.setTextColor(animation.animatedValue as Int)
    }

    companion object {

        private const val COLOR_CHANGE_DURATION = 1000
        private const val START_DELAY = 300
    }
}
